---
title: "DataAnalysis"
author: "Robyn Samuel"
date: "`r Sys.Date()`"
output: html_document
---

Install packages
```{r}
options(repos = c(CRAN = "https://cloud.r-project.org"))
install.packages("devtools")
library(devtools)
source("https://raw.githubusercontent.com/joey711/phyloseq/master/inst/scripts/installer.R",
       local = TRUE)
install_phyloseq(branch = "devel")
library(phyloseq)
if (!requireNamespace("devtools", quietly = TRUE)){install.packages("devtools")}
devtools::install_github("jbisanz/qiime2R")
library(qiime2R)
install.packages("tidyverse")
library(tidyverse)
install.packages("BiocManager")
library(BiocManager)
BiocManager::install("microbiome") # this took a very long time
library(microbiome)
BiocManager::install("microbiome") # this took a very long time
library(microbiome)
install.packages("vegan")
library(vegan)
install.packages("ggplot2")
library(ggplot2)
install.packages("plyr")
library(plyr)
library(dplyr)
devtools::install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
library(pairwiseAdonis)
```
set working directory, read in data, and create a phyloseq object
```{r}
setwd(dir="/Users/RMS1U18/Library/CloudStorage/OneDrive-UniversityofSouthampton/PhD/Preservatives/Bioinformatics/CLEAN/16S_RNA/")

# read in data
metadata<-read_q2metadata("metadata.txt") # must contain q2:types in second line
SVs<-read_qza("table.qza")$data
SVs <- as.data.frame(SVs)
tree <- read_qza("rooted-tree.qza")$data

#remove extra samples and blanks
drops <- c("ctrl-1", "ctrl-2","ctrl-3", "mock-4", "mock-12", "NAP-B") 
SVs <- SVs[ , !(names(SVs) %in% drops)]
metadata <- metadata %>% 
  filter(`Keep`=="Yes")

# load taxonomy
taxonomy<-read_qza("taxonomy.qza")$data %>% parse_taxonomy()

# change metadata SampleID into row.names
metadata2 <- metadata[,-1]
rownames(metadata2) <- metadata[,1]

# load data to phyloseq
OTU = otu_table(as.matrix(SVs), taxa_are_rows = TRUE)              
SAM = sample_data(metadata2, errorIfNULL = TRUE)                
TAX = tax_table(as.matrix(taxonomy, rownames = TRUE))
TRE = phy_tree(tree)
data_phylo <- phyloseq(OTU, TAX, SAM, TRE) 

# If your having issues loading the data into a phyloseq object use the following ven diagram to see if your sample names match the metadata.
# It might be an issue with rownames not being in the first column 
#gplots::venn(list(taxonomy=rownames(metadata.all2), featuretable=colnames(all.16S2)))
```
# Rarefy - weighted UniFrac - PCoA - PERMANOVA
Trying plysoseq tutorial for ordination: 
https://joey711.github.io/phyloseq/plot_ordination-examples.html
first rarefy the data and plot PCoA for weighted UniFrac
```{r}
theme_set(theme_bw())

#Remove OTUs that do not show appear more than 5 times in more than half the samples
GP = data_phylo # changed name to match tutorial

# Rarefy to even sampling depth.
GP1 = rarefy_even_depth(GP, sample.size = 36800, trimOTUs = TRUE, rngseed = TRUE) #uses default rngseed of 711
GP.ord <- ordinate(GP1, "PCoA", "unifrac", weighted=TRUE) 
p1 = plot_ordination(GP1, GP.ord, type="samples", color="Preservative") 
p1 + geom_polygon(aes(fill=Preservative), alpha=0.5) + geom_point(size=5) + ggtitle("16S_RNA")
ggsave("16S_RNA-PCoA-wUF-rarefy.pdf", height=4, width=6, device="pdf") # save a PDF 4 inches by 8 inches
```
now look at statistical difference with PERMANOVA and post hoc pairwise testing
```{r}
#For rarefied weighted Unifrac:

weighted_unifrac = UniFrac(physeq=GP1, weighted=T, normalized=T, parallel=T, fast=T)

vegan::adonis2(weighted_unifrac ~ phyloseq::sample_data(GP1)$Preservative)

pairwise.adonis(weighted_unifrac, phyloseq::sample_data(GP1)$Preservative)
```

# Proportions - Bray-Curtis - PCoA - PERMANOVA
now looking at the data normalised to proportions
```{r}
# normalize to proportions
GP2 = microbiome::transform(GP, "compositional")
GP.ord <- ordinate(GP2, "PCoA", "bray")
p2 = plot_ordination(GP2, GP.ord, type="samples", color="Preservative") 
p2 + geom_polygon(aes(fill=Preservative), alpha=0.5) + geom_point(size=5) + ggtitle("16S_RNA")
ggsave("16S_RNA-PcoA-bray-proportion.pdf", height=4, width=6, device="pdf") # save a PDF 4 inches by 8 inches
```
now looking statistical difference with PERMANOVA
```{r}

# For proportional Bray Curtis:

#ps_tr <- microbiome::transform(GP, "compositional")

ps_dist_matrix <- phyloseq::distance(GP2, method ="bray")

vegan::adonis2(ps_dist_matrix ~ phyloseq::sample_data(GP2)$Preservative)

pairwise.adonis(ps_dist_matrix, phyloseq::sample_data(GP2)$Preservative)




```
# Alpha diversity
now looking at alpha diversity with phyloseq tutorial https://joey711.github.io/phyloseq/plot_richness-examples.html
```{r}
theme_set(theme_bw())
pal = "Set1"

# trimming any ASVs that are not in any samples (they might have been in blanks or any other samples that wer removed)
GP <- prune_taxa(taxa_sums(data_phylo) > 0, data_phylo)

plot_richness(GP)
plot_richness(GP, measures=c("Observed", "Shannon", "Simpson"))
p <- plot_richness(GP, x="Preservative", color="Preservative",  measures=c("Observed", "Shannon", "Simpson"))
p + geom_boxplot(aes(fill=Preservative), alpha= 0.5) #alpha is to make the fill 50% transparent
ggsave("16S_RNA-alpha-diversity.pdf", height=3, width=8, device="pdf") # save a PDF 4 inches by 8 inches
```

save the alpha diversity measures to the sample data and perform non parametric statistical analysis for each meausre
```{r}
sample_data(GP)$observed.physeq <- estimate_richness(GP, measures="Observed")
sample_data(GP)$shannon.physeq <- estimate_richness(GP, measures="Shannon")
sample_data(GP)$simpson.physeq <- estimate_richness(GP, measures="Simpson")

# export metadata from phyloseq
diversity <- data.frame(GP@sam_data)

# statistical test for difference (non-normal distribution)
# Observed
kruskal.test(observed.physeq$Observed ~ Preservative, data = diversity)
pairwise.wilcox.test(diversity$observed.physeq$Observed, diversity$Preservative, p.adjust.method="fdr")
# Shannon
kruskal.test(shannon.physeq$Shannon ~ Preservative, data = diversity)
pairwise.wilcox.test(diversity$shannon.physeq$Shannon, diversity$Preservative, p.adjust.method="fdr")
# Simpson
kruskal.test(simpson.physeq$Simpson ~ Preservative, data = diversity)
pairwise.wilcox.test(diversity$simpson.physeq$Simpson, diversity$Preservative, p.adjust.method="fdr")

```
